src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,157 @@
+
+    /**
+     * Returns the default BufferRecyclerPool implementation which is the thread local based one.
+     */
+    static BufferRecyclerPool defaultRecyclerPool() {
+        return PoolStrategy.THREAD_LOCAL.getPool();
+    }
+
+    enum PoolStrategy {
+        NO_OP(new NonRecyclingRecyclerPool()),
+        THREAD_LOCAL(new ThreadLocalRecyclerPool()), 
+        CONCURRENT_DEQUEUE(new ConcurrentDequePool()),
+        LOCK_FREE(new LockFreePool());
+
+        private final BufferRecyclerPool pool;
+
+        PoolStrategy(BufferRecyclerPool pool) {
+            this.pool = pool;
+        }
+
+        public BufferRecyclerPool getPool() {
+            return pool;
+        }
+    }
+
+    /**
+     * Default {@link BufferRecyclerPool} implementation that uses
+     * {@link ThreadLocal} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ThreadLocalRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return BufferRecyclers.getBufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     *
+     * @since 2.16
+     */
+    class NonRecyclingRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ConcurrentDequePool implements BufferRecyclerPool {
+        private final Deque<BufferRecycler> pool = new ConcurrentLinkedDeque<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
+        }
+
+        private BufferRecycler getBufferRecycler() {
+            BufferRecycler bufferRecycler = pool.pollFirst();
+            return bufferRecycler != null ? bufferRecycler : new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+            pool.offerLast(bufferRecycler);
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * a lock free linked list for recycling instances.
+     *
+     * @since 2.16
+     */
+    class LockFreePool implements BufferRecyclerPool {
```

Comment:
I'd ask for maximum size, but that might not be possible with lock-free implementation?

Reply:
The same things that I wrote in [this other comment](https://github.com/FasterXML/jackson-core/pull/1064#discussion_r1307382381) are also valid here. That said, please let me know if you also want a bounded implementation of the pool (that shouldn't be based on a linked list/deque anyway) and I will add it.

Reply:
@mariofusco I think I would like a bounded variant too, if it's not overly complicated (tons of code). Having unbounded alternatives is fine then, too, with brief note in comments.